README.txt
Demonstrates how to use CouchDB with your Extension.

Get your free CouchDB for testing purposes on www.couchone.com

This Extension uses libraries from the PHP on Couch Project https://github.com/dready92/PHP-on-Couch

For Bugs, Suggestions and other stuff please visit this project in TYPO3 forge http://forge.typo3.org/projects/extension-relax


ChangeLog
2010-12-23 Ingo Pfennigstorf  <i.pfennigstorf@gmail.com>
	* Ajax Calls to read Documents
	* #11642 jQuery has to be loaded in any way - no dependency at the moment
	* #11643 Added JS Code
	* Release of Version 0.1.0

2010-12-22 Ingo Pfennigstorf  <i.pfennigstorf@gmail.com>
	* Added possibility to configure the Databasename via TypoScript
	* Using TYPO3 functions to instantiate Classes
	* Dummy Class to pass all Attributes to
	* Check if DB exists
	* Output Changes
	* #11639 Read from the couch

2010-12-20 Ingo Pfennigstorf  <i.pfennigstorf@gmail.com>
	* Writing to CouchDB
	* Defining default fields for a special purpose (May be changed via TypoScript)
	* Cleaned up code

2010-12-19 Ingo Pfennigstorf  <i.pfennigstorf@gmail.com>
	* Added PHP Couch Library

2010-12-18 Ingo Pfennigstorf  <i.pfennigstorf@gmail.com>
	* Creating Extension


pub/js/relax.js
$(document).ready(function(){

	/*Dynamically add some fields*/
	$('.addmore').click(function(){

		var rfn = generateRandomFieldName();
		$(this).append('<br /> <input type="text" name="tx-relax-pi1['+rfn+'][fieldname]" id="tx-relax-pi1-'+rfn+'" />'+
			'	<input type="text" name="tx-relax-pi1['+rfn+'][fieldvalue]" id="tx-relax-pi1-'+rfn+'[value]" /> ');

	})


	/*Posting Stuff to an Ajax function*/
	$('.document').click(function() {

		$.post('/index.php?eID=relax', {
			docID:$(this).parent().attr('id')
		}, function(data) {

			$('#doctext').css('display','block').html(data);

		});

		return false;
	});

});


function generateRandomFieldName(){
	var chars = "0123456789abcdefghiklmnopqrstuvwxyz";
	var string_length = 8;
	var randomstring = '';
	for (var i=0; i<string_length; i++) {
		var rnum = Math.floor(Math.random() * chars.length);
		randomstring += chars.substring(rnum,rnum+1);
	}

	return randomstring;
}

lib/Document.php
<?php

/* * *************************************************************
 *  Copyright notice
 *
 *  (c) 2010 Ingo Pfennigstorf <i.pfennigstorf@gmail.com>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 * ************************************************************* */

/**
 * Dummy Document - no further Programming required
 * @author ingop
 */
class tx_relax_lib_Document {

}

?>


pub/tx_relax_eid.php
<?php

tslib_eidtools::connectDB();
tslib_eidtools::initFeUser();

// initialize TSFE
require_once(PATH_tslib . 'class.tslib_fe.php');
require_once(PATH_t3lib . 'class.t3lib_page.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couch.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couchDocument.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couchClient.php');

$temp_TSFEclassName = t3lib_div::makeInstanceClassName('tslib_fe');
$GLOBALS['TSFE'] = new $temp_TSFEclassName($TYPO3_CONF_VARS, $pid, 0, true);
$GLOBALS['TSFE']->connectToDB();
$GLOBALS['TSFE']->initFEuser();
$GLOBALS['TSFE']->determineId();
$GLOBALS['TSFE']->getCompressedTCarray();
$GLOBALS['TSFE']->initTemplate();
$GLOBALS['TSFE']->getConfigArray();

$abbrev = $GLOBALS['TSFE']->tmpl->setup['plugin.']['tx_relax_pi1.']['couch.'];
$couchHost = $abbrev['host'];
$couchPort = $abbrev['port'];
$couchDbName = $abbrev['dbName'];

$couchDsn = $couchHost . ':' . $couchPort . '/';

$docId = $_POST['docID'];

$db = t3lib_div::makeInstance('couchClient', $couchDsn, $couchDbName);
try {
	$doc = $db->getDoc($docId);
} catch (Exception $e) {
	if ($e->getCode() == 404) {
		echo "The chosen Document does not exist";
	}
	exit(1);
}
$doc = couchDocument::getInstance($db, $docId);

echo "<h3>Data from your Document</h3>";
getSavedData($doc);

//display data
function getSavedData($doc) {

	foreach ($doc->getFields() as $key => $val) {
		echo $key . ' - ' . $val . '<br />';
	}
}

?>

pi1/class.tx_relax_pi1.php
<?php

/* * *************************************************************
 *  Copyright notice
 *
 *  (c) 2010 Ingo Pfennigstorf <ingo.pfennigstorf@gmail.com>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 * ************************************************************* */
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */
require_once(PATH_tslib . 'class.tslib_pibase.php');
require_once(PATH_typo3conf . '/ext/relax/lib/Document.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couch.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couchClient.php');
require_once(PATH_typo3conf . '/ext/relax/lib/couchDocument.php');

/**
 * Plugin 'Relax and take it easy' for the 'relax' extension.
 *
 * @author	Ingo Pfennigstorf <ingo.pfennigstorf@fh-brandenburg.de>
 * @package	TYPO3
 * @subpackage	tx_relax
 */
class tx_relax_pi1 extends tslib_pibase {

	var $prefixId = 'tx_relax_pi1';  // Same as class name
	var $scriptRelPath = 'pi1/class.tx_relax_pi1.php'; // Path to this script relative to the extension dir.
	var $extKey = 'relax';
	var $pi_checkCHash = false;
	/**
	 * Couch Options
	 */
	private $couchDsn;
	private $couch;
	private $defaultFields;
	private $couchDB;
	/**
	 *
	 * Avalable Databses
	 * @var array
	 */
	private $couches;

	/**
	 * The main method of the PlugIn
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	function main($content, $conf) {
		$this->conf = $conf;
		$this->pi_setPiVarDefaults();
		$this->pi_loadLL();
		$this->pi_USER_INT_obj = 1;

		$GLOBALS["TSFE"]->additionalHeaderData[$this->extKey] = '<script type="text/javascript" src="/typo3conf/ext/relax/pub/js/relax.js"></script>';

		//get default fields From TYPO3 Conf
		$this->defaultFields = $this->getDefaultFields($conf['defaultFields']);

		//Database selected via TypoScript
		$this->couchDB = $conf['couch.']['dbName'];

		//Compute DSN
		$this->couchDsn = $conf['couch.']['host'] . ':' . $conf['couch.']['port'] . '/';

		//Start Connection to CouchDB Server
		$this->couch = $this->initDB();

		//check if we have some parameters
		$gp = t3lib_div::_GP('tx-relax-pi1');

		if ($gp['add']) {

			$gp = $this->optimiereFelder($gp);
			$content .= $this->addToCouch($gp);
		}

		if ($gp['docId'] && !$gp['add']) {
			$content .= $this->showDocument($gp['docId']);
		}

		//Get all DBs and load into an Array
		$this->couches = $this->getAvailableDBs();

		//Comparing defined DB with available
		$this->compareAvailableSelected();

		//Add the HTML Form to Content
		$content .= $this->addForm();

		$content .= $this->readCouch();

		return $this->pi_wrapInBaseClass($content);
	}

	private function showDocument($docId) {

	}

	/**
	 * Get all Document IDs from our DB
	 * @return <type>
	 */
	private function readCouch() {

		$documents = $this->couch->getAllDocs();

		$c = '<ul>';
		foreach ($documents->rows as $document) {
			$c .= '<li id="' . $document->id . '" >' . $this->getDocLink($document->id) . '</li>';
		}
		$c .='</ul>';

		return $c;
	}

	private function getDocLink($docId) {

		$conf = array(
			'parameter' => $GLOBALS['TSFE']->id . ' _self document',
			'additionalParams' => '&tx-relax-pi1[docId]=' . $docId,
//			'ATagParams' => 'class="document"'
		);

		$link = $this->cObj->typoLink($docId, $conf);

		return $link;
	}

	/**
	 * Optimizing the Key-Value-Pairs to fit
	 * Removing Tempfields to Key/Value pairs
	 * @param array $gp
	 * @return array
	 */
	private function optimiereFelder($gp) {
		$data = array();
		foreach ($gp as $key => $value) {

			if (is_array($value)) {
				$key = $value['fieldname'];
				$value = $value['fieldvalue'];
			}

			$key != '' ? $data[$key] = $value : null;
		}
		return $data;
	}

	/**
	 *
	 * Get default Fields From TypoScript
	 * @param string $felder
	 */
	private function getDefaultFields($felder) {
		$defaultFelder = explode(',', $felder);

		$fields = array();
		foreach ($defaultFelder as $feld) {
			$fields[] = trim($feld);
		}

		return $fields;
	}

	/**
	 * Add Data to Your couchDB Server
	 * @return string Message from CouchDB Server
	 */
	private function addToCouch($daten) {

		//Instantiate Dummy Document
		$doc = t3lib_div::makeInstance('tx_relax_lib_Document');

		//Pass all attributes to the document
		foreach ($daten as $key => $val) {
			$doc->$key = $val;
		}

		//and try to store this document to our CouchDB
		try {
			$nachricht = $this->couch->storeDoc($doc);
		} catch (Exception $e) {
			$nachricht = $e->getTraceAsString();
		}

		return $nachricht;
	}

	/**
	 * Initializing Connection
	 * @return couchClient
	 */
	private function initDB() {

		$db = t3lib_div::makeInstance('couchClient', $this->couchDsn, $this->couchDB);

		if ($db) {
			return $db;
		} else {
			throw new Exception('Connection to CouchDB failed ...', 4151351342344);
		}
	}

	/**
	 * Yeah
	 * Generating the Form
	 * @TODO Extract to Template
	 * @TODO Deliver Javascript
	 */
	private function addForm() {
		$html .= '
		<div id="doctext" style="height:300px;display:none"> </div>
		<form name="couchfields" action="' . $this->pi_getPageLink($GLOBALS['TSFE']->id) . '" method="post">';
		foreach ($this->defaultFields as $field) {


			$html .= '
			<label for="tx-relax-pi1-' . $field . '">' . ucfirst($field) . '</label>
			<input type="text" name="tx-relax-pi1[' . $field . ']" id="tx-relax-pi1-' . $field . '" />
				<br />';
		}
		$tempField = $this->getRandomFieldName();
		$html .= '
		
			<input type="text" name="tx-relax-pi1[' . $tempField . '][fieldname]" id="tx-relax-pi1-' . $tempField . '[label]" />
			<input type="text" name="tx-relax-pi1[' . $tempField . '][fieldvalue]" id="tx-relax-pi1-' . $tempField . '[value]" />
			<div class="addmore">+</div>
			<br />
			<input type="submit" name="tx-relax-pi1[add]" value="add" />
			
		</form>
		';

		return $html;
	}

	/**
	 * Generates a random String
	 * @return string Random String
	 */
	private function getRandomFieldName() {
		return t3lib_div::shortMD5(time());
	}

	/**
	 * Show all Databases in your Couch Instance
	 */
	private function getAvailableDBs() {

		$databases = $this->couch->listDatabases();
		return $databases;
	}

	/**
	 * Check if the configured DB is in the list of availables
	 * @return boolean
	 */
	private function compareAvailableSelected() {
		if (in_array($this->couchDB, $this->couches)) {
			return true;
		} else {
			throw new Exception("Selected Database not available", 900843510546, $previous);
		}
	}

}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/relax/pi1/class.tx_relax_pi1.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/relax/pi1/class.tx_relax_pi1.php']);
}
?>

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "relax".
#
# Auto generated 19-12-2010 13:54
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'CouchDB Document Storage',
	'description' => 'Stores all your data in a CouchDB - without taking care of a relational database model!',
	'category' => 'fe',
	'author' => 'Ingo Pfennigstorf',
	'author_email' => 'ingo.pfennigstorf@gmail.com',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 1,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.0.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:22:{s:9:"ChangeLog";s:4:"29db";s:10:"README.txt";s:4:"ee2d";s:12:"ext_icon.gif";s:4:"dd89";s:17:"ext_localconf.php";s:4:"c214";s:14:"ext_tables.php";s:4:"ab1e";s:11:"favicon.ico";s:4:"2ab2";s:13:"locallang.xml";s:4:"3a66";s:16:"locallang_db.xml";s:4:"42a4";s:19:"doc/wizard_form.dat";s:4:"c794";s:20:"doc/wizard_form.html";s:4:"fefc";s:13:"lib/couch.php";s:4:"15d6";s:18:"lib/couchAdmin.php";s:4:"80cc";s:19:"lib/couchClient.php";s:4:"09e8";s:21:"lib/couchDocument.php";s:4:"2d4c";s:23:"lib/couchReplicator.php";s:4:"5566";s:14:"pi1/ce_wiz.gif";s:4:"02b6";s:26:"pi1/class.tx_relax_pi1.php";s:4:"b616";s:34:"pi1/class.tx_relax_pi1_wizicon.php";s:4:"8d48";s:13:"pi1/clear.gif";s:4:"cc11";s:17:"pi1/locallang.xml";s:4:"15f0";s:30:"pi1/static/relax/constants.txt";s:4:"d41d";s:26:"pi1/static/relax/setup.txt";s:4:"99ee";}',
	'suggests' => array(
	),
);

?>